tests/test/misc.rs

```diff
@@ -812,8 +812,10 @@ fn env_bound_vars() {
                     WellFormed(&'a ())
                 }
             }
-        } yields {
+        } yields[SolverChoice::slg_default()] {
             expect![["Ambiguous; definite substitution for<?U0> { [?0 := '^0.0] }"]]
+        } yields[SolverChoice::recursive_default()] {
+            expect![[r#"Unique; for<?U0> { substitution [?0 := '^0.0] }"#]]
```

Comment:
So, which is more "correct"? Unique or Ambiguous?

Reply:
Unique, I think. I'm not sure why the SLG solver is getting ambiguous.

Reply:
Let me post some more details!

Reply:
OK, so, the SLG solver's aggregate function is too conservative and could use this same optimization:

https://github.com/rust-lang/chalk/blob/4b95a894b8395f670f959e224d5f0ad58a642e8c/chalk-engine/src/slg/aggregate.rs#L26-L32

Right now it assumes that if there are two answers, it should return ambiguous, but that's not true if one of those answers is a strict subset of the other. In this case we get two answers:

* `?0 = '^0.0` with a region constraint; and
* `?0 = '^0.0` without a region constraint

The latter is the "unique answer" that the recursive solver returns, because it is strictly better than the first one.

Reply:
OK, I pushed a semi-hacky fix for that.

Reply:
I reverted this semi-hacky-fix because I didn't like it. Going to mark this *concern* as resolved though.